docs/DeveloperGuide.md

```diff
@@ -1,38 +1,726 @@
+Implementation of `parseUserInput`:
 
+![CommandParser implementation](diagrams/CommandParserSequence.png)
+
+`parseUserInput(String userInput)` is used to directly convert a string into their argument-payload pairs.
+It first calls `splitIntoCommand` to split input over the `' --'` delimiter to get a `String[] commands`.
+Each `command` in `commands[]` contain the argument and payload. Internally, it splits the
+argument from the payload and populates a `HashMap` with the one-to-one mapping. After all `command`s have
+been processed, the map is returned to the `Manager` for usage.
+
+**Sample Code**
+
+```java
+// Example usage to get the HashMap
+public class FooManager extends Manager {
+    public HashMap<String, String> handleCommand(String userInput) {
+        // Get a reference to the parser
+        CommandParser parser = getCommandParser();
+        // Get the one-to-one mapping
+        HashMap<String, String> result = parseUserInput(userInput);
+        return result;
+    }
+}
+```
+
+**Usage: `getMainCommand(userInput)`**
+
+To understand what the user wants to do, we need a convenient way to get the `mainCommand` from the user input.
+The canonical way to do this is to use `getMainCommand`. This defeats adversarial input where the main command
+is input as an argument.
+
+Internally, this just splits the string by whitespace and returns the first word in the array.
+
+#### Alternative Designs Considered
+
+We considered alternative command structures such as [AB3](https://se-education.org/addressbook-level3/UserGuide.html)
+where input types are
+specified , `e.g. n/John Doe` which more 'secure' from the get go.
+However, due to the following issues, AB3 was not chosen as the alternative solution compared to the shell-like
+structure.
+
+**Steep learning curve**  
+For experienced and inexperienced users, it is a hassle to remember what letter corresponds to what argument.
+For AB3, the user needs to remember all the different `char` 'verbs' such as `e/` for email, `n/` for name.
+This violates design consideration (1).
+
+**Does not scale well**  
+AB3 structure runs the high risk of argument-space collision as well.  
+For example, consider a command that needs an "email" and "entry". What does `e/<payload>` correspond to?
+We could simply just put entry as *some other character* -- but that defeats the purpose of having the structure in the
+first place as the character is the argument's first character.
+This makes behaviour **unpredictable** and a **confusing** user experience.
+
+**Bad expert user experience**
+
+For expert users and CLI-masters, pedantic argument input like AB3 makes the typing experience MUCH slower due to the
+need to type which is relatively clunky as the user will need to type far off to the '/' key on the keyboard.
+
+### AtomicHabit Component
+
+![AtomicHabit Component](diagrams/AtomicHabit.png)
```

Comment:
To simplify the diagram, maybe not every command has to be shown, which would definitely reduce the amount of arrows